torchtune/utils/ppo/_generation.py

```diff
@@ -0,0 +1,185 @@
+    *,
+    max_generated_tokens: int,
+    pad_id: int = 0,
+    temperature: float = 1.0,
+    top_k=None,
+    stop_tokens=None,
+    custom_generate_next_token=None,
+    dtype: torch.dtype = torch.float32,
+):
+    """
+    Generates tokens from a model conditioned on a prompt. In contrast to ~torchtune.utils._generation.generate~,
+    this function handles input sequences which may be left-padded by shifting position IDs
+    and correctly masking out padding tokens.
+
+    Args:
+        model (TransformerDecoder): model used for generation
+        prompt (torch.Tensor): tensor with the token IDs associated with the given prompt,
+            with shape either [seq_length] or [bsz x seq_length]. Padded sequences should be
+            left-padded with `pad_id` for sequence collation.
+        max_generated_tokens (int): number of tokens to be generated
+        pad_id (int): token ID to use for padding, default 0.
+        temperature (float): value to scale the predicted logits by, default 1.0.
+        top_k (Optional[int]): If specified, we prune the sampling to only token ids within the top_k probabilities,
+            default None.
+        stop_tokens (Optional[List[int]]): If specified, generation is stopped when any of these tokens are generated,
+            default None.
+        custom_generate_next_token (Optional[Callable]): If specified, we'll use the ``custom_generate_next_token function``.
+            This is generally only useful if you want to specify a ``torch.compile`` version of the generate next token for
+            performance reasons. If None, we use the default ``generate_next_token`` function. Default is None.
+
+    Examples:
+        >>> model = torchtune.models.llama3.llama3_8b()
+        >>> tokenizer = torchtune.models.llama3.llama3_tokenizer()
+        >>> prompt = [0, 0, 0] + tokenizer("Hi my name is") # substitute 0 with pad_id
+        >>> output = generate(model, torch.tensor(prompt), max_generated_tokens=100)
+        >>> print(tokenizer.decode(output[0]))
+        ?? ?? ?? Hi my name is Jeremy and I'm a friendly language model assistant!
+
+    Returns:
+        List[List[int]]: collection of lists of generated tokens
+    """
+    prompt = prompt.view(1, -1) if prompt.ndim == 1 else prompt
+
+    bsz, prompt_length = prompt.size()
+    generated_tokens = prompt.clone()
+    generated_tokens[generated_tokens == pad_id] = 0
+
+    if stop_tokens is not None:
+        # convert stop tokens to tensor for easy matching
+        stop_tokens = (
+            torch.tensor(stop_tokens, device=prompt.device) if stop_tokens else None
+        )
+        # keeps track at a high level if we've already hit a stop token in a sequence so we can early stop
+        stop_token_reached = torch.zeros(bsz, dtype=torch.bool, device=prompt.device)
+        # everything in stop_token_mask starts as 1s, and we'll set them to 0 for sequences
+        # that already hit a stop token
+        stop_token_mask = torch.ones(
```

Comment:
I don't actually need to support stop tokens here atm.

Reply:
Saying that, a more recent implementation of RLHF [from OpenAI](https://github.com/openai/summarize-from-feedback/blob/8af822a428c93432aa80ffbe5b065a8f93895669/summarize_from_feedback/utils/experiment_helpers.py#L19) does stop sampling when hitting EOS tokens. I'll need to re-work things slightly to make things work with truncation during generation, since I use the truncation masks for rejection sampling, and I use the attention masks and position IDs created when generating during PPO optimisation steps.